Show RPS histogram in valkey-benchmark

hanxizh9910 opened the pull request:
When benchmarking with the a target thoughput using the `--rps` flag, display an RPS histogram in the benchmark report. This can help identify if there is a bottleneck.

Related to #2219.

zuiderkwast commented:
@artikell

sarthakaggarwal97 commented:
> My idea is to use a histogram to statistically analyze the Requests Per Second (RPS) and provide a specific distribution, such as p50, p90, and p99 values. Then, if the water level does not meet expectations, a prompt will be given.

@artikell I didn't completely get this. Would you please elaborate

hanxizh9910 commented:
> > My idea is to use a histogram to statistically analyze the Requests Per Second (RPS) and provide a specific distribution, such as p50, p90, and p99 values. Then, if the water level does not meet expectations, a prompt will be given.
> 
> @artikell I didn't completely get this. Would you please elaborate

I don't know if my understanding is correct or not: my understanding is that we should create a statistical distribution of RPS measurements and calculate percentiles (p50, p90, p99). We'll establish a water level threshold (e.g., 90% of the user's target RPS). For each percentile, we check if it meets or exceeds this threshold. If any percentile falls below the water level, we then identify whether the bottleneck is on the client side (benchmark) or server side.

artikell commented:
> > > My idea is to use a histogram to statistically analyze the Requests Per Second (RPS) and provide a specific distribution, such as p50, p90, and p99 values. Then, if the water level does not meet expectations, a prompt will be given.
> > 
> > 
> > @artikell I didn't completely get this. Would you please elaborate
> 
> I don't know if my understanding is correct or not: my understanding is that we should create a statistical distribution of RPS measurements and calculate percentiles (p50, p90, p99). We'll establish a water level threshold (e.g., 90% of the user's target RPS). For each percentile, we check if it meets or exceeds this threshold. If any percentile falls below the water level, we then identify whether the bottleneck is on the client side (benchmark) or server side. Could you please you provide more details @artikell ?

Here's the thing. We want users to know that the benchmark has reached a bottleneck. Locating the specific bottleneck is quite difficult, and I believe it's not within the scope of the benchmark's responsibility.

hanxizh9910 committed:
Add RPS bottleneck detection when actual RPS < 90% of target

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

hanxizh9910 committed:
Fix clang-format issues in RPS bottleneck detection

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

hanxizh9910 committed:
valkey-benchmark: Add RPS bottleneck detection with dual thresholds

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

hanxizh9910 committed:
Removed the trailing whitespace to pass clang-format test

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

hanxizh9910 commented:
> There is a difference between p99 and 90% here. If 10% of the requests do not meet expectations, it is also possible that the RPS has reached 90% @hanxizh9910

Thank you for the feedback! 

I've updated the implementation:

**New approach:**
- **≥99%**: "GOOD PERFORMANCE" - Successfully achieved target RPS
- **90-99%**: "BOTTLENECK DETECTED" - Performance below optimal target  
- **<90%**: "SEVERE BOTTLENECK DETECTED" - Performance significantly below target
<img width="484" height="111" alt="Screenshot 2025-08-19 at 11 03 38 AM" src="https://github.com/user-attachments/assets/a676f5e7-632c-4b6d-a065-0b5550dd687b" />
<img width="457" height="94" alt="Screenshot 2025-08-19 at 11 03 56 AM" src="https://github.com/user-attachments/assets/fe9665cf-b5f8-4f3b-847f-0916d32d1358" />
<img width="472" height="110" alt="Screenshot 2025-08-19 at 11 04 49 AM" src="https://github.com/user-attachments/assets/7a3cc5b9-fb37-492b-908d-24346ac3a426" />

Would you mind taking a look at this updated approach? @artikell

artikell commented:
@hanxizh9910 I didn't give a good example before, but here we need to use percentiles instead of averages. To avoid misjudgment.

hanxizh9910 commented:
> @hanxizh9910 I didn't give a good example before, but here we need to use percentiles instead of averages. To avoid misjudgment.

@artikell Thank you for your clarification. I have a question about implementing RPS percentiles for short benchmark runs.

The issue: When benchmarks complete quickly, we get very few RPS samples:

Examples:

valkey-benchmark -n 1000 --rps 1000 → completes in ~1 second → only 1 RPS sample

valkey-benchmark -n 3000 --rps 1000 → completes in ~3 seconds → only 3 RPS samples

### Potential solution - Dynamic sampling intervals:
Instead of fixed 1-second intervals, could we use adaptive intervals to always get a consistent number of samples?

Approach: Always divide benchmark duration into 100 samples

- 5-second benchmark: Sample every 50ms (5s ÷ 100) → 100 samples

  - Expected requests per 50ms interval: 1000 RPS × 0.05s = 50 requests
  
- 30-second benchmark: Sample every 300ms (30s ÷ 100) → 100 samples

  - Expected requests per 300ms interval: 1000 RPS × 0.3s = 300 requests

- 2-minute benchmark: Sample every 1.2s (120s ÷ 100) → 100 samples

  - Expected requests per 1.2s interval: 1000 RPS × 1.2s = 1200 requests

This would ensure we always get exactly 100 RPS samples regardless of benchmark duration, making percentile calculations consistent and meaningful.

### Questions:

Do you think 100 samples is a good target number for reliable percentile analysis?

Or would you prefer a different approach for handling variable benchmark durations?

hanxizh9910 committed:
benchmark: Add RPS percentile analysis for bottleneck detection

Adds 1-second RPS sampling and percentile-based performance guarantees
to improve bottleneck detection when using --rps parameter.

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

hanxizh9910 committed:
Fix clang-format issues in RPS percentile implementation

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

hanxizh9910 committed:
benchmark: Fix code formatting issues

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

hanxizh9910 commented:
> @hanxizh9910 The meaning of RPS is the meaning of every second, so we can use second level intervals. And we guarantee at least 1 second of data. Just make sure there is one sample.
> 
> For data that does not meet the requirement of 1 second, such as 200ms data with 1.2 seconds, this type of data can be skipped without affecting statistics.
> 
> Most users are more concerned about whether there is insufficient pressure during stress testing.

Thanks for the great suggestion! I've implemented RPS percentile analysis based on your feedback. Here's what I added:

New functionality:

- 1-second RPS sampling : Measures actual RPS every 1000ms during benchmark execution
- HDR histogram tracking : Uses same histogram library as latency analysis for consistent percentile calculation
- Performance guarantees display : Shows user-friendly "X% of time: RPS ≥ Y" instead of confusing percentile values

Example output: 

<img width="361" height="147" alt="Screenshot 2025-08-28 at 11 22 46 AM" src="https://github.com/user-attachments/assets/2d6b2e13-e702-4c28-aa07-051aeaa149c8" />
<img width="370" height="142" alt="Screenshot 2025-08-28 at 11 20 21 AM" src="https://github.com/user-attachments/assets/5c4e329f-28b5-462e-b173-e2af3037113f" />
<img width="464" height="217" alt="Screenshot 2025-08-28 at 11 16 08 AM" src="https://github.com/user-attachments/assets/14ecfe3e-4b1b-4f46-b1fc-6433ac13284f" />

For tests shorter than 1 second, it falls back to the original average-based analysis to avoid misleading percentile data:

<img width="309" height="62" alt="Screenshot 2025-08-28 at 2 01 05 PM" src="https://github.com/user-attachments/assets/e74e8867-014e-43d9-9bde-2412c31ef7fd" />


Would appreciate your feedback! @artikell

sarthakaggarwal97 commented on src/valkey-benchmark.c:
```diff
@@ -1160,6 +1163,55 @@ static void showLatencyReport(void) {
+            printf("    Target RPS: %.0f\n", target_rps);
+            printf("    Actual RPS: %.2f (%.1f%% of target)\n", reqpersec, achievement_pct);
+
+            // ADD RPS PERCENTILE ANALYSIS
```
```suggestion
            /* ADD RPS PERCENTILE ANALYSIS */
```

sarthakaggarwal97 commented on src/valkey-benchmark.c:
```diff
@@ -1160,6 +1163,55 @@ static void showLatencyReport(void) {
+                const float rps_p1 = hdr_value_at_percentile(config.rps_histogram, 1.0);
+                const float rps_p10 = hdr_value_at_percentile(config.rps_histogram, 10.0);
+                const float rps_p50 = hdr_value_at_percentile(config.rps_histogram, 50.0);
+                // const float rps_p90 = hdr_value_at_percentile(config.rps_histogram, 90.0);
```
Can we remove these comments if they are unused?

sarthakaggarwal97 commented on src/valkey-benchmark.c:
```diff
@@ -1160,6 +1163,55 @@ static void showLatencyReport(void) {
+                    printf("    Performance significantly below target.\n");
+                }
+            } else {
+                // Fallback to original average-based analysis
```
let's fix formatting style for all the comments.

sarthakaggarwal97 commented on src/valkey-benchmark.c:
```diff
@@ -1224,6 +1276,15 @@ static void benchmarkSequence(const char *title, char *cmd, int len, int seqlen)
 
+    if (config.rps > 0) {
+        hdr_init(1,       // Min RPS: 1
+                 1000000, // Max RPS: 1M
```
Is there a limit to what valkey benchmark can support?

zuiderkwast replied:
We can do more than 1M with threading + pipelining + fast machine.

sarthakaggarwal97 reviewed (changes requested):
Took an initial look. Will post more feedback.

hanxizh9910 committed:
Modified the comment style

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

zuiderkwast commented on src/valkey-benchmark.c:
```diff
@@ -1160,6 +1163,55 @@ static void showLatencyReport(void) {
+
+                // Bottleneck analysis using P1 (99% of time achieved this or better)
+                if (rps_p1 >= 0.99 * target_rps) {
+                    printf("    Status: GOOD PERFORMANCE\n");
```
I'm not sure if no bottleneck always means good performance. If we tested with a very low RPS, we don't know much.

sarthakaggarwal97 replied:
yeah, it was in my pending comments too. I don't think we need to print this, we can just detect if the bottleneck is observed.

sarthakaggarwal97 commented on src/valkey-benchmark.c:
```diff
@@ -1924,6 +1986,30 @@ int parseOptions(int argc, char **argv) {
     exit(exit_status);
 }
 
+static void sampleRPS(void) {
```
I am not sure about the use of this function. Can we use similar mechanics as latency to build the histogram, they already do that to generate percentile values. I am guessing we also just need to build the right histogram for RPS and then let the user know if we were able to reach that RPS or not.

hanxizh9910 committed:
Removed the print statement for good performance

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

hanxizh9910 committed:
Removed trailing spaces to pass clang-format-check

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

hanxizh9910 committed:
Fix: replace the fixed max RPS to a dynamic max

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

hanxizh9910 committed:
Modified the format

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

hanxizh9910 committed:
Used instantaneous_rps to calculate the percentile

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

hanxizh9910 committed:
Removed an unused function call

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

hanxizh9910 committed:
Fixed a casting problem

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

hanxizh9910 committed:
Improve RPS bottleneck detection using p99

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

sarthakaggarwal97 commented on src/valkey-benchmark.c:
```diff
@@ -1955,6 +1989,11 @@ long long showThroughput(struct aeEventLoop *eventLoop, long long id, void *clie
     const float instantaneous_dt = (float)(current_tick - config.previous_tick) / 1000.0;
     const float instantaneous_rps = (float)(requests_finished - previous_requests_finished) / instantaneous_dt;
+
+    if (config.rps_histogram) {
```
did you try running the benchmarks with threads option? Can we check if it's running as expected with multiple threads and clients?

hanxizh9910 replied:
yes, i did. It's running as expected
<img width="649" height="695" alt="Screenshot 2025-10-03 at 12 49 06 PM" src="https://github.com/user-attachments/assets/b685aed5-a82e-4439-b7b5-187029bb9f18" />

zuiderkwast replied:
Should we print the RPS percentiles below the latency percentiles, above the summary section?

Is the RPS percentiles useful for anything other than detecting a bottleneck? I'm still thinking we may want to hide it if no bottleneck is detected, to avoid spamming/bloating the output, if/when it is not useful.

hanxizh9910 replied:
Hi, i just updated the code based on your suggestion, now the RPS section will only be shown when there is a bottleneck. It will also appear above the summary section:

<img width="458" height="681" alt="Screenshot 2025-10-06 at 12 20 28 PM" src="https://github.com/user-attachments/assets/d6d4cd6d-2586-4818-93eb-852c4ae79570" />
<img width="462" height="662" alt="Screenshot 2025-10-06 at 12 20 39 PM" src="https://github.com/user-attachments/assets/9f629521-3174-4a93-a27c-649b1b7eb553" />

hanxizh9910 committed:
Moved the showrps above summary section and only show it if there is bottleneck

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

sarthakaggarwal97 commented on src/valkey-benchmark.c:
```diff
@@ -1076,6 +1077,35 @@ static void createMissingClients(client c) {
+        float target_rps = (float)config.rps;
+        const char *rps_status = NULL;
+
+        if (p99 < 0.9 * target_rps) {
```
This logic is a little confusing one, but I think it means that if 99% of the requests are less than 90% of the throughput, then we are severely bottlenecked. If 99% of the requests are less than RPS, then it is just bottlenecked.

I would suggest to make it P90 instead of P99 for both cases (as it is too lenient). If 90% of the requests are not able to reach 90% of the set RPS, it is a bottleneck. I also feel these thresholds 90% of the throughput, and 100% of the throughput, to detect bottleneck, could be a configuration that a user can set. Not everyone may agree with our thresholds.

@zuiderkwast what do you think?

sarthakaggarwal97 replied:
@hanxizh9910 when the logic is finalized, it would be nice to add a comment to explain this as well.

zuiderkwast replied:
I'm don't really know. I think @artikell is better to answer this question.

artikell replied:
Yes, it is appropriate to set this up as a configuration here. Indeed, not everyone accepts 90%.

zuiderkwast replied:
More configuration? I'm skeptical to more configuration. It's hard to understand for users.

If we are uncertain, why not just print the info in that case so the user can judge from the output?

hanxizh9910 replied:
Hi @zuiderkwast, I added a print statement to explain how the bottleneck is detected. Does it look good to you?
<img width="569" height="394" alt="Screenshot 2025-10-09 at 12 53 45 PM" src="https://github.com/user-attachments/assets/c0d71891-3a6e-4534-b3b5-193ad2256d1b" />

sarthakaggarwal97 replied:
> More configuration? I'm skeptical to more configuration. It's hard to understand for users.
If we are uncertain, why not just print the info in that case so the user can judge from the output?

I like that @zuiderkwast always tries to avoid new configurations 😄. I think it's a much better idea that should print say the %age of requests that didn't meet the RPS configurations and let the user judge it. I'd prefer if we don't make the decisions on behalf of the user. @artikell do you have any objections to this?

artikell replied:
LTGM.

hanxizh9910 replied:
> > More configuration? I'm skeptical to more configuration. It's hard to understand for users.
> > If we are uncertain, why not just print the info in that case so the user can judge from the output?
> 
> I like that @zuiderkwast always tries to avoid new configurations 😄. I think it's a much better idea that should print say the %age of requests that didn't meet the RPS configurations and let the user judge it. I'd prefer if we don't make the decisions on behalf of the user. @artikell do you have any objections to this?

Hi Sarthak @sarthakaggarwal97 , thank you for your feedback. I modified the code. Now, the output looks like this: 
<img width="483" height="646" alt="Screenshot 2025-10-13 at 11 26 51 AM" src="https://github.com/user-attachments/assets/93ce109f-8828-43e5-af50-34cd622d1e4c" />
If the user specifies the rps parameter, the RPS summary will be displayed; otherwise, only the latency will be shown:
<img width="475" height="629" alt="Screenshot 2025-10-13 at 11 27 09 AM" src="https://github.com/user-attachments/assets/5372b085-789f-488c-815e-dfaa4daac5f0" />

sarthakaggarwal97 commented on src/valkey-benchmark.c:
```diff
@@ -1076,6 +1077,35 @@ static void createMissingClients(client c) {
+        }
+
+        if (rps_status) {
+            printf("\nRPS Percentiles:\n");
```
can we use the similar printing format as latency?

sarthakaggarwal97 reviewed (changes requested):
The PR looks better, left some comments to discuss.

hanxizh9910 committed:
Changed the format of RPS summary to align with that of the latency summary

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

hanxizh9910 committed:
Fixed a format issue

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

hanxizh9910 committed:
Added info telling the user how do we detect bottleneck

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

hanxizh9910 committed:
Show the RPSSummary all the time

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

zuiderkwast reviewed (commented):
After the last changes, it looks like the code is now ignoring the `-q` (quiet) and `--csv` options. Did you test with these? For example like this:

    ./valkey-benchmark -q --rps 1000 -t set
    ./valkey-benchmark --csv --rps 1000 -t set

Also I suggested earlier that we print the RPS percentiles above the Summary section and only when there is a bottleneck and you changed according to these suggestions. Now, it's changed again to be printed below the summary and always, without an explanation. I can accept changes if there is an explanation. :)

I still think the Summary should be printed last. The Summary includes the throughput (RPS). It's not only about latencies.

        printf("Summary:\n");
        printf("  throughput summary: %.2f requests per second\n", reqpersec);
        printf("  latency summary (msec):\n");
        printf("    %9s %9s %9s %9s %9s %9s\n", "avg", "min", "p50", "p95", "p99", "max");
        printf("    %9.3f %9.3f %9.3f %9.3f %9.3f %9.3f\n", avg, p0, p50, p95, p99, p100);

hanxizh9910 committed:
Printing RPS Summarybefore the latency distribution

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

hanxizh9910 committed:
Only print the RPS Info when there are 10% or more samples below 90% * target rps

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

hanxizh9910 committed:
Changed the showLatencyReport function name to showReport

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

zuiderkwast commented:
> We were thinking that if we cannot decide a default threshold for what counts as a bottleneck, we could simply print the RPS section whenever the user uses the --rps option.

I understand. OK, I'm fine with that, if you prefer it.

hanxizh9910 committed:
Always printing the RPS Summary

Signed-off-by: Hanxi Zhang <hanxizh@amazon.com>

hanxizh9910 commented:
@zuiderkwast thank you for your feedback! This is the latest output:
<img width="481" height="647" alt="Screenshot 2025-10-14 at 3 41 05 PM" src="https://github.com/user-attachments/assets/c8626b6c-4065-4e9c-aa53-707563a8934b" />
It will always print the RPS summary if the user uses the --rps option

sarthakaggarwal97 reviewed (approved):
LGTM! I had discussed offline with @hanxizh9910 but forgot to approve it :D
Thanks @hanxizh9910 for taking up the change.

sarthakaggarwal97 commented:
yeah @zuiderkwast thanks for updating the description, that is important.
